### PR TITLE
GC old blocks / chunks

### PR DESCRIPTION
**Pull request by Kouprin**
Fixes https://github.com/nearprotocol/nearcore/issues/1998.

### CODE REVIEW

**Comment by Kouprin** [comment]
https://gitlab.com/near-protocol/nearcore/-/jobs/414674604
test/providers.test.js failed because it asks for GCed block at height 1

**Comment by Kouprin** [comment]
Created https://github.com/nearprotocol/nearlib/issues/199.

File: chain/client/src/client.rs [hunk 1]
```diff
@@ -705,6 +705,12 @@ impl Client {
 
         if status.is_new_head() {
             self.shards_mgr.update_largest_seen_height(block.header.inner_lite.height);
+            match self.chain.clear_old_data() {
```
Comment (bowenwang1996) [inline chain/client/src/client.rs]: `if let Err(err) = self.chain.clear_old_data()`

File: chain/chain/src/store.rs [hunk 2]
```diff
@@ -2292,4 +2360,57 @@ mod tests {
         assert_ne!(block_hash, block_hash1);
         assert_ne!(epoch_id_to_hash, epoch_id_to_hash1);
     }
+
+    #[test]
```
Comment (bowenwang1996) [inline chain/chain/src/store.rs]: Can you also test that `clear_old_data` works for multiple heights?

File: chain/chain/src/chain.rs [hunk 3]
```diff
@@ -501,6 +507,24 @@ impl Chain {
         });
     }
 
+    pub fn clear_old_data(&mut self) -> Result<(), Error> {
+        let mut chain_store_update = self.store.store_update();
+        let head = chain_store_update.head()?;
+        if head.height >= NUM_EPOCHS_TO_KEEP_STORE_DATA * self.epoch_length {
```
[the remark this answers is not in the document]
Comment (Kouprin) [reply chain/chain/src/chain.rs]: ++

File: chain/chain/src/store.rs [hunk 4]
```diff
@@ -243,14 +240,13 @@ pub trait ChainStoreAccess {
         &mut self,
         hash: &CryptoHash,
     ) -> Result<&LightClientBlockView, Error>;
-    /// Check if we have block header at given height across any chain.
     /// Returns a hashmap of epoch id -> block hash that we can use to determine whether the block is double signed
     /// For each epoch id we need to store just one block hash because for the same epoch id the signer of a given
     /// height must be the same.
-    fn get_any_block_hash_by_height(
+    fn get_all_block_hashes_by_height(
```
[the remark this answers is not in the document]
Comment (Kouprin) [reply chain/chain/src/store.rs]: ++

File: chain/chain/src/store.rs [hunk 5]
```diff
@@ -1744,34 +1821,28 @@ impl<'a> ChainStoreUpdate<'a> {
...
         }
         for (hash, block) in self.chain_store_cache_update.blocks.iter() {
+            let mut map = match self
+                .chain_store
+                .get_all_block_hashes_by_height(block.header.inner_lite.height)
+            {
+                Ok(m) => m.clone(),
+                Err(_) => HashMap::new(),
+            };
+            if !map.contains_key(&block.header.inner_lite.epoch_id) {
```
[the remark this answers is not in the document]
Comment (Kouprin) [reply chain/chain/src/store.rs]: awesome
++

File: chain/client/src/client.rs (hunk 1 above)
Comment (Kouprin) [reply chain/client/src/client.rs]: ++

File: chain/chain/src/store.rs (hunk 2 above)
Comment (Kouprin) [reply chain/chain/src/store.rs]: ++

**Comment by Kouprin** [comment]
test/providers.test.js in nearlib fails, needs to be redone

**Pull request merged**